### PR TITLE
Fix volume symlink check

### LINKED ISSUE #11862: Volume and symbolic links in 1.6.0-rc2
`<mylife>`
I accidentally discover through docker-compose trying to give access to `/var/run/docker.sock` from inside a container
`</mylife>`

## Environment

Docker is running on Ubuntu 14.04 using the default aufs storage driver.

The goal is to create a volume in a directory which is actually a symbolic link. I use the `/var/run` directory which is a symlink to `/run` in Ubuntu 14.04

The test command will be:

    docker run --rm -ti --volume /var/run/gordon ubuntu /bin/bash

If `/var/run/gordon` does not exist on the host, the command is working find on both 1.5.0 and 1.6.0-rc2.

Now if I create a file on the host:

    sudo touch /var/run/gordon
    docker run --rm -ti --volume /var/run/gordon ubuntu /bin/bash

In 1.5.0 everything is going smoothly and a `/var/run/gordon` directory is created inside the container.

In 1.6.0-rc2, the container's creation is failing with the following error message:

    FATA[0000] Error response from daemon: file exists at %!s(MISSING), can't create volume there

The issue does not happen if I'm using `--volume /run/gordon`

### PR DESCRIPTION
**Pull request by cpuguy83**
Fixes #11862
carry of #11866

This fixes a seemingly obscure but bad issue caused when introducing volume path checking to make sure the volume path is not a file. (#10112)

The issue only seems obscure but I was able to hit it with my current setup with 1.6-rc2 with a compose config as such:

``` yaml
volumes:
    - "/var/run/docker.sock:/var/run/docker.sock"
```

I think because compose is first creating a container with a volume path of `/var/run/docker.sock` and then only applies the bind mount on start (instead of creating the bind directly), and since `/var/run` is a symlink to `/run` in debian based distros...
This triggers the path checking (as in the above mentioned PR) on create and never gets to the actual start.

Basically the issue is this:

Create a container with a symlink which points to a path (within the container) that matches an existing path on the host which is a file (not a dir).

### CODE REVIEW

**Comment by ggtools** [comment]
You can also create it without compose by using `docker run --volume /var/run/docker.sock` as explained in #11862

**Comment by cpuguy83** [comment]
Good point. I was really showing that to show how ubiquitous this error would be

**Comment by jfrazelle** [comment]
the test fails on Master and the Release branch so that is good

**Comment by jfrazelle** [comment]
LGTM ping @crosbymichael

**Comment by jfrazelle** [comment]
or @icecrime

**Comment by jfrazelle** [comment]
windows is failing on stupid pullVerified fyi

**Comment by jfrazelle** [comment]
wait NOT LGTM fails on lxc
```
--- FAIL: TestRunCreateVolumesInSymlinkDir (3.29s)
	docker_utils.go:604: "run -v /test/test test-volume-symlink" failed with errors: lxc-start: conf.c: mount_entry: 2049 Not a directory - failed to mount '/var/lib/docker/vfs/dir/717d0a8409d806b4a9c37493c072a0024733b4d6b856422b2e8e3e3f2cce47dd' on '/usr/local/lib/lxc/rootfs///test/test'
		lxc-start: conf.c: lxc_setup: 4163 failed to setup the mount entries for 'c5d72d8a8ab6a66d15f0b4963ae65d48331ee4a0185caa8308d8f0e3c853e2bf'
		lxc-start: start.c: do_start: 688 failed to setup the container
		lxc-start: sync.c: __sync_wait: 51 invalid sequence number 1. expected 2
		lxc-start: start.c: __lxc_start: 1080 failed to spawn 'c5d72d8a8ab6a66d15f0b4963ae65d48331ee4a0185caa8308d8f0e3c853e2bf'
		lxc-start: lxc_start.c: main: 342 The container failed to start.
		lxc-start: lxc_start.c: main: 346 Additional information can be obtained by setting the --logfile and --logpriority options.
```

**Comment by jfrazelle** [comment]
so either skip it or... idk maybe lxc does not support symlinks ping @ashahab-altiscale

File: integration-cli/docker_cli_run_test.go [hunk 1]
```diff
@@ -477,6 +477,37 @@ func TestRunWithVolumesFromExited(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	f.Close()
+
+	dockerFile := fmt.Sprintf("FROM busybox\nRUN mkdir -p %s\nRUN ln -s %s /test", dir, dir)
+	if _, err := buildImage(name, dockerFile, false); err != nil {
+		t.Fatal(err)
+	}
+
```
Comment (coolljt0725) [inline integration-cli/docker_cli_run_test.go]: I think we should have a `defer deleteImages(name)` :)

Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Yep, thanks!

**Comment by ashahab-altiscale** [comment]
Yes, lxc does support symlinks, just not in the same layer: https://github.com/docker/docker/issues/10466

This will pass:
```
03:16:52-ashahab~$ docker build -t test1 - << END
FROM busybox
RUN mkdir -p /tmp
RUN ln -s /tmp /test
END

Sending build context to Docker daemon 2.048 kB
Sending build context to Docker daemon 
Step 0 : FROM busybox
 ---> 4986bf8c1536
Step 1 : RUN mkdir -p /tmp
 ---> Running in 1dd85ae51c09
 ---> 226c244ee43b
Removing intermediate container 1dd85ae51c09
Step 2 : RUN ln -s /tmp /test
 ---> Running in 33b5eb4aeabd
 ---> 531e8094d8cc
Removing intermediate container 33b5eb4aeabd
Successfully built 531e8094d8cc
```
But this will fail:
```
03:17:08-ashahab~$ docker build -t test1 - << END
FROM busybox
RUN mkdir -p /tmp1
RUN ln -s /tmp1 /test
END

Sending build context to Docker daemon 2.048 kB
Sending build context to Docker daemon 
Step 0 : FROM busybox
 ---> 4986bf8c1536
Step 1 : RUN mkdir -p /tmp1
 ---> Running in b58dae41bab2
 ---> fcfc0315105e
Removing intermediate container b58dae41bab2
Step 2 : RUN ln -s /tmp1 /test
 ---> Running in 417f75434876
INFO[0002] open /proc/2084/stat: no such file or directory 
```
Can you break your directory creation and symlinking into 2 layers?

**Comment by cpuguy83** [comment]
@ashahab-altiscale It is in two layers, but I guess I also don't understand why your first example works and your 2nd one fails.

**Comment by ashahab-altiscale** [comment]
Doesn't docker build squeeze everything into one layer?

**Comment by ashahab-altiscale** [comment]
Oh the first passes because /tmp already existed, was in a different layer.

**Comment by cpuguy83** [comment]
No, each build step is a separate layer.

**Comment by jfrazelle** [comment]
@cpuguy83 we need this in the release :( but I dont want to break the lxc build

**Comment by jfrazelle** [comment]
what if we skip it

**Comment by cpuguy83** [comment]
Updated to skip the test on LXC since it's specifically an issue with the test implementation on LXC and not the actual fix.

If someone can suggest a way to make sure that test works on LXC that'd be great too.

**Comment by jfrazelle** [comment]
It'd be a good issue for someone to play with :)

**Comment by ashahab-altiscale** [comment]
Create the dir in a previous docker build statement. Or symlink an existing
dir.

**Comment by cpuguy83** [comment]
@ashahab-altiscale But that's exactly what's happening.
It's:
```Dockerfile
FROM busybox
RUN mkdir -p <dir>
RUN ln -s <dir> /test
```

**Comment by ashahab-altiscale** [comment]
@cpuguy83 I updated my issue.
Can you try this with relative paths for everything? At least relative path for the symlink?

**Comment by tiborvass** [comment]
LGTM

File: integration-cli/docker_cli_run_test.go [hunk 2]
```diff
@@ -477,6 +477,39 @@ func TestRunWithVolumesFromExited(t *testing.T) {
...
 
+// Volume path is a symlink which also exists on the host, and the host side is a file not a dir
+// But the volume call is just a normal volume, not a bind mount
+func TestRunCreateVolumesInSymlinkDir(t *testing.T) {
+	testRequires(t, SameHostDaemon)
+	testRequires(t, NativeExecDriver)
+	defer deleteAllContainers()
+	name := "test-volume-symlink"
+
+	dir, err := ioutil.TempDir("", name)
```
Comment (ashahab-altiscale) [inline integration-cli/docker_cli_run_test.go]: Instead of creating a tempdir, can you provide it a relative dir name?

**Comment by jfrazelle** [comment]
can we fanagle with it in a follow up PR, you can assign to me I don't mind

File: integration-cli/docker_cli_run_test.go (hunk 2 above)
Comment (cpuguy83) [inline integration-cli/docker_cli_run_test.go]: Hmm, I think this would not trigger the error case.

**Comment by jfrazelle** [comment]
LGTM
